lib/core/src/restore.rs

```diff
@@ -0,0 +1,1172 @@
+                &tx_map,
+                histories.send_chain,
+                &immutable_db.send_chain_swap_immutable_db_by_swap_id,
+            )
+            .await?;
+        let recovered_chain_receive_data = self
+            .recover_receive_chain_swap_tx_ids(
+                &tx_map,
+                histories.receive_chain,
+                &immutable_db.receive_chain_swap_immutable_db_by_swap_id,
+            )
+            .await?;
+
+        Ok(RecoveredOnchainData {
+            send: recovered_send_data,
+            receive: recovered_receive_data,
+            chain_send: recovered_chain_send_data,
+            chain_receive: recovered_chain_receive_data,
+        })
+    }
+
+    /// Reconstruct Send Swap tx IDs from the onchain data and the immutable DB data
+    async fn recover_send_swap_tx_ids(
+        &self,
+        tx_map: &TxMap,
+        send_histories_by_swap_id: HashMap<String, SendSwapHistory>,
+    ) -> Result<HashMap<String, RecoveredOnchainDataSend>> {
+        let mut res: HashMap<String, RecoveredOnchainDataSend> = HashMap::new();
+        for (swap_id, history) in send_histories_by_swap_id {
+            // If a history tx is one of our outgoing txs, it's a lockup tx
+            let lockup_tx_id = history
+                .iter()
+                .find(|&tx| tx_map.outgoing_tx_map.contains_key::<Txid>(&tx.txid))
+                .cloned();
+            if lockup_tx_id.is_none() {
+                error!("No lockup tx found when recovering data for Send Swap {swap_id}");
+            }
+
+            // If a history tx is one of our incoming txs, it's a refund tx
+            let refund_tx_id = history
+                .iter()
+                .find(|&tx| tx_map.incoming_tx_map.contains_key::<Txid>(&tx.txid))
+                .cloned();
+
+            // A history tx that is neither a known incoming or outgoing tx is a claim tx
+            let claim_tx_id = history
+                .iter()
+                .filter(|&tx| !tx_map.incoming_tx_map.contains_key::<Txid>(&tx.txid))
+                .find(|&tx| !tx_map.outgoing_tx_map.contains_key::<Txid>(&tx.txid))
+                .cloned();
+
+            res.insert(
+                swap_id,
+                RecoveredOnchainDataSend {
+                    lockup_tx_id,
+                    claim_tx_id,
+                    refund_tx_id,
+                },
+            );
+        }
+
+        Ok(res)
+    }
+
+    /// Reconstruct Receive Swap tx IDs from the onchain data and the immutable DB data
+    async fn recover_receive_swap_tx_ids(
+        &self,
+        tx_map: &TxMap,
+        receive_histories_by_swap_id: HashMap<String, ReceiveSwapHistory>,
+    ) -> Result<HashMap<String, RecoveredOnchainDataReceive>> {
+        let mut res: HashMap<String, RecoveredOnchainDataReceive> = HashMap::new();
+        for (swap_id, history) in receive_histories_by_swap_id {
+            let (lockup_tx_id, claim_tx_id) = match history.len() {
+                2 => {
```

Comment:
We should consider here <=2 as there might be only lockup but no claim.

Reply:
Done: c7575afbf96dcf34f1cc40026c7b21241a31c865